controllers/capabilities/tenant_controller.go

```diff
@@ -98,6 +97,61 @@ func (r *TenantReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
+	// Tenant has been marked for deletion
+	if tenantR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
+		existingTenant, err := controllerhelper.FetchTenant(tenantR, portaClient)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// delete tenantCR if tenant is present in 3scale
+		if existingTenant != nil {
+			// do not attempt to delete tenant that is already scheduled for deletion
+			if existingTenant.Signup.Account.State != scheduledForDeletionState {
+				err := controllerhelper.DeleteTenant(tenantR, portaClient)
+				if err != nil {
+					r.EventRecorder.Eventf(tenantR, corev1.EventTypeWarning, "Failed to delete tenant", "%v", err)
+					return ctrl.Result{}, err
+				}
+			} else {
+				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion %v", existingTenant.Signup.Account.ID)
+			}
+		}
+
+		// add or remove finalizer, tenant CR will be deleted if tenant in 3scale does not exists, if tenant is deleted or if it's already marked for deletion
+		err = controllerhelper.ReconcileFinalizers(tenantR, r.Client, tenantFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
+	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
+	if tenantR.GetDeletionTimestamp() != nil {
+		return ctrl.Result{}, nil
+	}
+
+	changed := tenantR.SetDefaults()
+	if changed {
+		err = r.Client.Update(context.TODO(), tenantR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		reqLogger.Info("Tenant resource updated with defaults")
+		// Expect for re-trigger
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
+		err = controllerhelper.ReconcileFinalizers(tenantR, r.Client, tenantFinalizer)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
```

Comment:
should be `return ctrl.Result{Requeue: true}, nil` or it will stop reconciling and the tenant will not be created

Reply:
That's not the case, the tenants are created and I think (correct me if this is wrong) it's because by editing the tenant CR the reconciler is triggered again, but setting requeue: true, we are going to have 2 runs, one after finalizer is added, and one coming from the update done to tenant CR

Reply:
When I create the tenant CR, the operator will get a reconciliation event. The controller will add the finalizer and stop the reconciliation job. No new reconciliation will be done until there is a update or delete event.  So, effectively, the tenant will not be created in 3scale until you edit the CR after having created that. 

Maybe you can test and tell me if I am wrong

Reply:
I did test it and it works as expect. The way I see it adding finalizer is also an update to CR